src/openstack_cloud/openstack_cloud.py

```diff
@@ -0,0 +1,606 @@
+                    "Attempting clean up of openstack server %s that timeout during creation",
+                    full_name,
+                )
+                try:
+                    conn.delete_server(name_or_id=full_name, wait=True)
+                except (
+                    openstack.exceptions.SDKException,
+                    openstack.exceptions.ResourceTimeout,
+                ):
+                    logger.exception(
+                        "Failed to cleanup openstack server %s that timeout during creation",
+                        full_name,
+                    )
+                self._delete_keypair(conn, instance_id)
+                raise OpenstackError(f"Timeout creating openstack server {full_name}") from err
+            except openstack.exceptions.SDKException as err:
+                logger.exception("Failed to create openstack server %s", full_name)
+                self._delete_keypair(conn, instance_id)
+                raise OpenstackError(f"Failed to create openstack server {full_name}") from err
+
+            return OpenstackInstance(server, self.prefix)
+
+    def get_instance(self, instance_id: str) -> OpenstackInstance | None:
+        """Get OpenStack instance by instance ID.
+
+        Args:
+            instance_id: The instance ID.
+
+        Returns:
+            The OpenStack instance if found.
+        """
+        full_name = self.get_server_name(instance_id)
+        logger.info("Getting openstack server with %s", full_name)
+
+        with _get_openstack_connection(
+            clouds_config=self._clouds_config, cloud=self._cloud
+        ) as conn:
+            server = OpenstackCloud._get_and_ensure_unique_server(conn, full_name)
+            if server is not None:
+                return OpenstackInstance(server, self.prefix)
+        return None
+
+    def delete_instance(self, instance_id: str) -> None:
+        """Delete a openstack instance.
+
+        Raises:
+            OpenstackError: Unable to delete OpenStack server.
+
+        Args:
+            instance_id: The instance ID of the instance to delete.
+        """
+        full_name = self.get_server_name(instance_id)
+        logger.info("Deleting openstack server with %s", full_name)
+
+        with _get_openstack_connection(
+            clouds_config=self._clouds_config, cloud=self._cloud
+        ) as conn:
+            try:
+                server = OpenstackCloud._get_and_ensure_unique_server(conn, full_name)
+                if server is not None:
+                    conn.delete_server(name_or_id=server.id)
+                OpenstackCloud._delete_keypair(conn, full_name)
+            except (
+                openstack.exceptions.SDKException,
+                openstack.exceptions.ResourceTimeout,
+            ) as err:
+                raise OpenstackError(f"Failed to remove openstack runner {full_name}") from err
+
+    def get_ssh_connection(self, instance: OpenstackInstance) -> SshConnection:
```

Comment:
Out of scope for this PR but just a reminder to myself to add a retry, perhaps using tenacity :D

Reply:
Currently, the callers of this method has retry (beside the delete runner, which will proceed to delete the runner without SSH to call the removal script).

We need to think about which layer we do the retry at.